packages/core/src/common/connection.ts

```diff
@@ -0,0 +1,270 @@
+// *****************************************************************************
+// Copyright (C) 2022 Ericsson and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { injectable } from 'inversify';
+import { Disposable, DisposableCollection } from './disposable';
+import { Emitter, Event } from './event';
+import { MaybePromise, serviceIdentifier } from './types';
+
+export enum ConnectionState {
+    OPENING,
+    OPENED,
+    CLOSING,
+    CLOSED,
+}
+export namespace ConnectionState {
+    /**
+     * @param connection
+     * @returns Whether `connection` is or is going to be closed.
+     */
+    export function isClosed<T>(connection: Connection<T>): connection is Connection<T> & { state: ConnectionState.CLOSING | ConnectionState.CLOSED } {
+        return connection.state === ConnectionState.CLOSING || connection.state === ConnectionState.CLOSED;
+    }
+}
+
+/**
+ * Helper type to get a `Connection<any>` without triggering the linter.
+ */
+export type AnyConnection = Connection<any>;
+
+/**
+ * A `Connection` allows you to listen for messages and send messages back.
+ *
+ * Messages must arrive integrally and in the order they are sent!
+ *
+ * Most implementations are going to be API adapters.
+ */
+export const Connection = serviceIdentifier<Connection<any>>('Connection');
+export interface Connection<T> {
+    readonly state: ConnectionState
+    onOpen: Event<void>
+    onClose: Event<void>
+    /**
+     * Emitted when something goes wrong with the underlying transport.
+     * Typically when a message fails to be sent or received.
+     */
+    onError: Event<Error>
+    onMessage: Event<T>
+    sendMessage(message: T): void
+    close(): void
+}
```

Comment:
Second most important abstraction: a normalized interface to send/receive messages.

We handle so many APIs to do communications (sockets, node IPC, electron IPC, custom multiplexed channels, etc) and none really has the same API. We use them all in the same way: to send and receive messages.

This type wouldn't be required if it weren't for the other components I created that are used to help with setting up communication stacks.

e.g. I need to do multiplexing over different IPC APIs for Electron but I only want one multiplexer implementation. By using the `Connection` type I can just write adapters for each API and use both in the same `DefaultConnectionMultiplexer` implementation.

Another important thing to keep in mind with connections: They may not only send strings or buffers. Most APIs support automatic serialization, and some of those implementations do a much better job than any other `JSON.stringify` we would call ourselves.

e.g. Socket.io will optimize the transmission of buffers sent through it. By thinking we should just send strings we called `JSON.stringify` on our structures before passing them to Socket.io, preventing it from properly sending our data because we passed it as a raw string.

The idea with connections then is to understand what the underlying transport support sending and stick to that, unless we have more specific needs. You may see more `Connection<any>` than `Connection<string>` in this PR.